File: codes/python/25.k-个一组翻转链表.py

class Solution:
    def reverseKGroup(self, head: ListNode, k: int) -> ListNode:
        if not head:
            return None
        # 区间 [a, b) 包含 k 个待反转元素
        a = b = head
        for i in range(k):
            # 不足 k 个，不需要反转，base case
            if not b:
                return head
            b = b.next
        # 反转前 k 个元素
        newHead = self.reverse(a, b)
        # 递归反转后续链表并连接起来
        a.next = self.reverseKGroup(b, k) 
        return newHead

    """ 反转区间 [a, b) 的元素，注意是左闭右开 """
    def reverse(self, a: ListNode, b: ListNode) -> ListNode: 
        pre, cur, nxt = None, a, a
        # while 终止的条件改一下就行了
        while cur != b:
            nxt = cur.next
            cur.next = pre
            pre = cur
            cur = nxt
        # 返回反转后的头结点
        return pre

File: codes/golang/25.k-个一组翻转链表.go

func reverseKGroup(head *ListNode, k int) *ListNode {
	if head == nil {
		return nil
	}

	a, b := head, head
	// 找到需要翻转的区间 [a, b)
	for i := 0; i < k; i++ {
		if b == nil {
			return head
		}
		b = b.Next
	}

	// 反转区间内的链表
	newHead := reverse(a, b)
	// 递归将后续链表的区间也翻转，然后再将它链接到新的区间内
	a.Next = reverseKGroup(b, k)

	return newHead
}

// 翻转区间 [a, b)，包头不包尾
func reverse(a, b *ListNode) *ListNode {
	// 初始化三个指针
	pre, cur, nxt := (*ListNode)(nil), a, a

	// 循环将当前节点指向前一个节点，然后将前一个节点和当前节点往后移动
	for cur != b {
		nxt = cur.Next
		cur.Next = pre
		pre, cur = cur, nxt
	}

	// 返回新的头结点
	return pre
}